File: core/src/main/java/no/elg/infiniteBootleg/world/Material.java
package no.elg.infiniteBootleg.world;

import com.badlogic.gdx.graphics.g2d.TextureRegion;
import com.google.common.base.Preconditions;
import java.lang.reflect.Constructor;
import java.lang.reflect.InvocationTargetException;
import kotlin.NotImplementedError;
import no.elg.infiniteBootleg.KAssets;
import no.elg.infiniteBootleg.Settings;
import no.elg.infiniteBootleg.items.ItemType;
import no.elg.infiniteBootleg.world.blocks.FallingBlock;
import no.elg.infiniteBootleg.world.blocks.TickingBlock;
import no.elg.infiniteBootleg.world.blocks.TntBlock;
import no.elg.infiniteBootleg.world.blocks.Torch;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

/**
 * @author Elg
 */
public enum Material {
  AIR(null, ItemType.AIR, 0f, false, false, false, false, true),
  STONE(1.5f),
  BRICK(2f),
  DIRT(1f),
  GRASS(0.8f),
  TNT(TntBlock.class, ItemType.BLOCK, 0.5f, true, true, true, true, false),
  SAND(FallingBlock.class, 1f),
  TORCH(Torch.class, ItemType.BLOCK, 0.1f, false, false, true, true, true),
  GLASS(null, ItemType.BLOCK, 0.1f, true, false, true, false, true),
  DOOR(null, ItemType.BLOCK, 1f, false, true, true, false, true),
  ;
  @Nullable private final Constructor<?> constructor;
  @Nullable private final Constructor<?> constructorProtoBuf;
  private final boolean solid;
  private final boolean blocksLight;
  private final boolean placable;
  private final float hardness;
  @Nullable private final TextureRegion texture;
  @NotNull private final ItemType itemType;
  private final boolean luminescent;
  private final boolean transparent;

  private static final Material[] VALUES = values();

  Material(float hardness) {
    this(null, hardness);
  }

  Material(@Nullable Class<?> impl, float hardness) {
    this(impl, ItemType.BLOCK, hardness, true, true, true, false, false);
  }

  /**
   * @param impl The implementation a block of this material must have
   * @param itemType
   * @param hardness
   * @param solid If objects can pass through this material
   * @param blocksLight If this material will block light
   * @param placable If a block of this material can be placed by a player
   * @param transparent
   */
  Material(
      @Nullable Class<?> impl,
      @NotNull ItemType itemType,
      float hardness,
      boolean solid,
      boolean blocksLight,
      boolean placable,
      boolean luminescent,
      boolean transparent) {
    this.itemType = itemType;
    this.luminescent = luminescent;
    this.transparent = transparent;
    if (impl != null) {
      if (itemType == ItemType.BLOCK) {
        Preconditions.checkArgument(
            Block.class.isAssignableFrom(impl),
            name() + " does not have " + Block.class.getSimpleName() + " as a super type");
        try {
          constructor =
              impl.getDeclaredConstructor(
                  World.class, Chunk.class, int.class, int.class, Material.class);
          constructorProtoBuf = null;
        } catch (NoSuchMethodException e) {
          throw new IllegalStateException(
              "There is no constructor of "
                  + impl.getSimpleName()
                  + " with the arguments World, Chunk, int, int, Material");
        }
      } else {
        constructor = null;
        constructorProtoBuf = null;
      }
    } else {
      constructor = null;
      constructorProtoBuf = null;
    }
    this.solid = solid;
    this.blocksLight = blocksLight;
    this.placable = placable;
    this.hardness = hardness;
    if (Settings.client && itemType != ItemType.AIR) {
      texture = KAssets.INSTANCE.getBlockAtlas().findRegion(name().toLowerCase());
      if (texture == null) {
        throw new NullPointerException("Failed to find a texture for " + name());
      }
    } else {
      texture = null;
    }
  }

  public static @NotNull Material fromOrdinal(int b) {
    return VALUES[b];
  }

  /**
   * @param world World this block this exists in
   * @param chunk
   * @param localX Relative x in the chunk
   * @param localY Relative y in the chunk
   * @return A block of this type
   */
  @NotNull
  public Block createBlock(@NotNull World world, @NotNull Chunk chunk, int localX, int localY) {
    Preconditions.checkArgument(isBlock());
    if (constructor == null) {
      return new BlockImpl(world, chunk, localX, localY, this);
    }
    try {
      return (Block) constructor.newInstance(world, chunk, localX, localY, this);
    } catch (InstantiationException | IllegalAccessException | InvocationTargetException e) {
      throw new IllegalStateException(e);
    }
  }

  public boolean isBlock() {
    return itemType == ItemType.BLOCK || itemType == ItemType.AIR;
  }

  public boolean create(@NotNull World world, int worldX, int worldY) {
    Material currentMaterial = world.getMaterial(worldX, worldY);
    if (currentMaterial != this && currentMaterial == AIR) {
      if (isBlock()) {

        Block block = world.setBlock(worldX, worldY, this, true, true);
        if (block instanceof TickingBlock tickingBlock) {
          tickingBlock.delayedShouldTick(1L);
        }
        return block != null;
      } else if (isEntity()) {
        throw new NotImplementedError("TODO :)");
      }
      throw new IllegalStateException(
          "This material (" + name() + ") is neither a block nor an entity");
    }
    return false;
  }

  public boolean isEntity() {
    return itemType == ItemType.ENTITY;
  }

  @Nullable
  public TextureRegion getTextureRegion() {
    return texture;
  }

  public boolean isSolid() {
    return solid;
  }

  public boolean isBlocksLight() {
    return blocksLight;
  }

  public boolean isPlacable() {
    return placable;
  }

  /**
   * @return If the texture of the material has any transparency
   */
  public boolean isTransparent() {
    return transparent;
  }

  public float getHardness() {
    return hardness;
  }

  /**
   * @return If this material emits light
   */
  public boolean isLuminescent() {
    return luminescent;
  }

  public boolean isSuperclassOf(Class<?> interfaze) {
    if (constructor == null) {
      return false;
    }
    return interfaze.isAssignableFrom(constructor.getDeclaringClass());
  }
}

File: core/src/main/kotlin/no/elg/infiniteBootleg/world/Material.kt
package no.elg.infiniteBootleg.world

import com.badlogic.ashley.core.Entity
import io.github.oshai.kotlinlogging.KotlinLogging
import no.elg.infiniteBootleg.Settings
import no.elg.infiniteBootleg.items.ItemType
import no.elg.infiniteBootleg.items.MaterialItem
import no.elg.infiniteBootleg.main.Main
import no.elg.infiniteBootleg.protobuf.ProtoWorld
import no.elg.infiniteBootleg.protobuf.material
import no.elg.infiniteBootleg.util.LocalCoord
import no.elg.infiniteBootleg.util.WorldCoord
import no.elg.infiniteBootleg.util.component1
import no.elg.infiniteBootleg.util.component2
import no.elg.infiniteBootleg.util.findTextures
import no.elg.infiniteBootleg.util.safeWith
import no.elg.infiniteBootleg.util.serverRotatableTextureRegion
import no.elg.infiniteBootleg.util.stringifyCompactLoc
import no.elg.infiniteBootleg.world.blocks.Block
import no.elg.infiniteBootleg.world.blocks.BlockImpl
import no.elg.infiniteBootleg.world.chunks.Chunk
import no.elg.infiniteBootleg.world.ecs.api.ProtoConverter
import no.elg.infiniteBootleg.world.ecs.components.ExplosiveComponent
import no.elg.infiniteBootleg.world.ecs.components.required.EntityTypeComponent.Companion.entityTypeComponent
import no.elg.infiniteBootleg.world.ecs.creation.createBlockEntity
import no.elg.infiniteBootleg.world.ecs.creation.createContainerEntity
import no.elg.infiniteBootleg.world.ecs.creation.createDoorBlockEntity
import no.elg.infiniteBootleg.world.ecs.creation.createGravityAffectedBlockEntity
import no.elg.infiniteBootleg.world.ecs.creation.createLeafEntity
import no.elg.infiniteBootleg.world.ecs.explosiveBlockFamily
import no.elg.infiniteBootleg.world.ecs.load
import no.elg.infiniteBootleg.world.render.texture.RotatableTextureRegion
import no.elg.infiniteBootleg.world.world.World
import java.util.concurrent.CompletableFuture

private val logger = KotlinLogging.logger {}

/**
 * @author Elg
 */
enum class Material(
  /**
   * How hard it is to break this material
   */
  val hardness: Float,
  val textureName: String? = null,
  /**
   * @return If the texture of the material has any transparency
   */
  val hasTransparentTexture: Boolean,
  /**
   * @return If this material can be collided with
   */
  val isCollidable: Boolean = true,
  /**
   * @return If this material blocks light
   */
  val blocksLight: Boolean = true,
  /**
   * @return If this material emits light
   */
  val emitsLight: Boolean = false,
  /**
   * @return If this material can be rotated, but the entity can handle the rendering
   */
  val invisibleBlock: Boolean = false,
  /**
   *
   * @return If this material can be handled by the player, otherwise this is a meta material
   */
  val canBeHandled: Boolean = true,
  private val createNew: ((world: World, chunk: Chunk, worldX: WorldCoord, worldY: WorldCoord, material: Material) -> CompletableFuture<Entity>)? = null
) : ContainerElement {
  AIR(
    hardness = 0f,
    hasTransparentTexture = true,
    isCollidable = false,
    blocksLight = false,
    emitsLight = false,
    invisibleBlock = true,
    canBeHandled = false
  ),
  STONE(hardness = 1.5f, hasTransparentTexture = false),
  BRICK(hardness = 2f, hasTransparentTexture = false),
  DIRT(
    hardness = 1f,
    hasTransparentTexture = false
  ),
  GRASS(hardness = 0.8f, hasTransparentTexture = false),
  TNT(
    hardness = 0.5f,
    hasTransparentTexture = false,
    createNew = { world: World, chunk: Chunk, worldX: WorldCoord, worldY: WorldCoord, material: Material ->
      world.engine.createBlockEntity(world, chunk, worldX, worldY, material, arrayOf(explosiveBlockFamily to "explosiveBlockFamily")) {
        safeWith { ExplosiveComponent() }
      }
    }
  ),
  SAND(hardness = 0.75f, hasTransparentTexture = false, createNew = { world: World, chunk: Chunk, worldX: WorldCoord, worldY: WorldCoord, material: Material ->
    world.engine.createGravityAffectedBlockEntity(world, chunk, worldX, worldY, material)
  }),
  TORCH(
    hardness = 0.1f,
    hasTransparentTexture = true,
    isCollidable = false,
    blocksLight = false,
    emitsLight = true,
    createNew = { world: World, chunk: Chunk, worldX: WorldCoord, worldY: WorldCoord, material: Material ->
      world.engine.createGravityAffectedBlockEntity(world, chunk, worldX, worldY, material)
    }
  ),
  GLASS(hardness = 0.1f, hasTransparentTexture = true, blocksLight = false),
  DOOR(
    hardness = 1f,
    hasTransparentTexture = true,
    isCollidable = false,
    blocksLight = false,
    invisibleBlock = true,
    createNew = { world: World, chunk, worldX: WorldCoord, worldY: WorldCoord, material: Material ->
      world.engine.createDoorBlockEntity(world, chunk, worldX, worldY, material)
    }
  ),
  BIRCH_TRUNK(hardness = 1.25f, hasTransparentTexture = true, isCollidable = false, blocksLight = false),
  BIRCH_LEAVES(
    hardness = 0.1f,
    hasTransparentTexture = true,
    isCollidable = false,
    blocksLight = false,
    createNew = { world: World, chunk, worldX: WorldCoord, worldY: WorldCoord, material: Material ->
      world.engine.createLeafEntity(world, chunk, worldX, worldY, material)
    }
  ),
  SANDSTONE(
    hardness = 1f,
    hasTransparentTexture = false
  ),
  CONTAINER(hardness = 1f, hasTransparentTexture = false, isCollidable = false, createNew = { world: World, chunk, worldX: WorldCoord, worldY: WorldCoord, material: Material ->
    world.engine.createContainerEntity(world, chunk, worldX, worldY, material)
  });

  override var textureRegion: RotatableTextureRegion? = null

  init {
    if (Settings.client) {
      textureRegion = this.findTextures(textureName)
      if (textureRegion == null && !canBeHandled) {
        throw NullPointerException("Failed to find a texture for $displayName")
      }
    } else {
      textureRegion = serverRotatableTextureRegion(textureName)
    }
  }

  /**
   * @param world World this block this exists in
   * @param chunk
   * @param localX Relative x in the chunk
   * @param localY Relative y in the chunk
   * @return A block of this type
   */
  fun createBlock(
    world: World,
    chunk: Chunk,
    localX: LocalCoord,
    localY: LocalCoord,
    protoEntity: ProtoWorld.Entity? = null
  ): Block {
    require(!chunk.isDisposed) { "Created block in disposed chunk" }
    return BlockImpl(chunk, localX, localY, this).also { block ->
      if (Main.isAuthoritative) {
        // Blocks client side should not have any entity in them
        val futureEntity = protoEntity?.let { world.load(it, chunk) } ?: createNew?.invoke(world, chunk, chunk.worldX + localX, chunk.worldY + localY, this)
        futureEntity?.thenApply { entity: Entity ->
          if (block.isDisposed) {
            world.removeEntity(entity)
            // This will fire when generating features in the world (i.e., trees next to other trees)
            logger.debug {
              "Block@${stringifyCompactLoc(block)} chunk ${stringifyCompactLoc(chunk)} was disposed" +
                " before entity (type ${entity.entityTypeComponent.hudDebug()}) was fully created. " +
                "Is the chunk disposed? ${chunk.isDisposed}"
            }
          } else {
            block.entity = entity
          }
        }
      }
    }
  }

  fun createBlocks(world: World, locs: LongArray, prioritize: Boolean = true, allowOverwiteNonAir: Boolean = false) {
    createBlocks(world, locs.asIterable(), prioritize, allowOverwiteNonAir)
  }

  fun createBlocks(world: World, locs: Iterable<Long>, prioritize: Boolean = true, allowOverwiteNonAir: Boolean = false) {
    val chunks = mutableSetOf<Chunk>()
    for ((worldX, worldY) in locs) {
      if (allowOverwiteNonAir || world.isAirBlock(worldX, worldY, markerIsAir = false)) {
        val block = world.setBlock(worldX, worldY, this, false, prioritize)
        chunks += block?.chunk ?: continue
      }
    }
    for (chunk in chunks) {
      chunk.updateTexture(prioritize)
    }
  }

  override val itemType: ItemType get() = ItemType.BLOCK

  override fun toItem(maxStock: UInt, stock: UInt): MaterialItem = MaterialItem(this, maxStock, stock)

  companion object : ProtoConverter<Material, ProtoWorld.Material> {

    /**
     * All materials that can be interacted in a normal fashion by the player
     */
    val normalMaterials: List<Material> = entries.filter(Material::canBeHandled)

    override fun Material.asProto(): ProtoWorld.Material =
      material {
        ordinal = this@asProto.ordinal
      }

    override fun ProtoWorld.Material.fromProto(): Material = Material.entries[ordinal]

    fun fromOrdinal(ordinal: Int): Material = entries[ordinal]
  }
}
